onnx/defs/training/defs.cc

```diff
@@ -0,0 +1,311 @@
+            "intermediate variables) that can be generated from inputs "
+            "cannot be included in this attribute.",
+            AttributeProto::STRINGS,
+            OPTIONAL)
+        .Attr(
+            "y",
+            "The targeted tensor. It can be viewed as the output of the "
+            "differentiated function. The attribute \"xs\" and attribute "
+            "\"zs\" are the minimal independent variable set that determines "
+            "the value of \"y\".",
+            AttributeProto::STRING)
+        .TypeConstraint(
+            "T1",
+            OpSchema::all_tensor_types(),
+            "Allow outputs to be any kind of tensor.")
+        .TypeConstraint(
+            "T2",
+            {"tensor(float16)",
+             "tensor(float)",
+             "tensor(double)"},
+            "Allow inputs to be any kind of floating-point tensor."));
+
+
+static const char* GraphCall_ver1_doc = R"DOC(
+The GraphCall operator invokes a graph inside TrainingInfoProto's
+algorithm field. The GraphCall inputs and outputs are bound to those of
+invoked graph by position. If a graph input has an initializer, that input
+is considered optional. All graph outputs are optional.
+
+Assume that ModelProto's graph field has
+- name: "MyInferenceGraph"
+- inputs: ["X", "W", "Z"]
+- outputs: ["Y"]
+
+with a model-level field
+
+- global_initializer: [W]
+
+as visualized below.
+
+```
+X -----.
+       |
+       v
+W --> Conv --> H --> Gemm --> Y
+                      ^
+                      |
+                      Z
+```
+
+Assume that the training algorithm contains
+
+- inputs: ["X_1", "Z_1", "C"]
+- outputs: ["W_new"]
+
+Inside the training algorithm graph, one can invoke the graph via adding
+a GraphCall node with
+
+- inputs: ["X_1", "W", Z_1"]
+- outputs: ["Y_1"]
+- an attribute graph_name="MyInferenceGraph",
+
+A possible algorithm graph may contain something like
+
+```
+.-------- W (declared in the model's global initializer list) 
+|   .-----'-----------.
+|   |                 v
+|   | .-- X_1 --> GraphCall(graph_name="MyInferenceGraph")
+|   | |            |  |
+|   | |            |  |
+|   | |   Z_1 -----'  |
+|   | |    |          V
+|   | |    |         Y_1 ---> Loss ---> O
+|   | |    |                    ^
+|   | |    |                    |
+|   | `--. |                    C
+|   |    | |                    |
+|   |    | |   .----------------'
+|   |    | |   |
+|   |    v v   v
+|   `--> Gradient(xs=["W"], zs=["X_1", "Z_1", "C"], y="O")
+|        |
+|        v
+|      dO_dW (gradient of W)
+|        |
+|        v
+`-----> Sub ----> W_new
+```
+
+where Loss is a dummy node which computes the minimized objective function.
+
+The variable "W" is an optional input in the called graph.
+If the user omits it, the input names of GraphCall becomes ["X_1", "", "Z_1"].
+In this case, from the view of computation graph, the Conv operator invoked by
+GraphCall's may be connected the global "W" variable.
```

Comment:
So when you use the gradient operator in the training graph, on a graph containing a `GraphCall`:
* Global initializers preserve gradient flow even when used by the inference graph but not explicitly passed to the `GraphCall`
* Inference graph initializers aren't visible to the training graph so you can't take the gradient w.r.t. them
* If the inference graph has an input with the same name as a global initializer, then that input is optional

Is that right?

I'm guessing the main purpose in adding the global intializers is to make it clearer that these are implicitly used by `GraphCall` without the `GraphCall` having to explicitly pass values for each one?

Reply:
> So when you use the gradient operator in the training graph, on a graph containing a `GraphCall`:
> 
> * Global initializers preserve gradient flow even when used by the inference graph but not explicitly passed to the `GraphCall`

Yes. Global variable is referenced if we don't find a local variable with the required name (here the name is "W").

> * Inference graph initializers aren't visible to the training graph so you can't take the gradient w.r.t. them

Yes. They are inference-time and training-time constants.

> * If the inference graph has an input with the same name as a global initializer, then that input is optional

Yes.

> 
> Is that right?
> 
> I'm guessing the main purpose in adding the global intializers is to make it clearer that these are implicitly used by `GraphCall` without the `GraphCall` having to explicitly pass values for each one?

Yes!